In [1]:
import os
import pandas as pd
import utils.DataUtils as DataUtils
import importlib

In [2]:
# 1. Get working folder
working_folder = DataUtils.get_current_folder()
print("Working folder:")
print(working_folder)

# 2. Get parser for removing code comments
parser = DataUtils.init_parser(working_folder=working_folder, list_languages=['rust', 'kotlin'])

Working folder:
/Users/tungthai/Library/CloudStorage/OneDrive-Personal/AdelaideStudy/2023/Trimester2/Project A/git/LLM_SV/Vulnerability Prediction/trimester3


In [17]:
def show_information(dataset_name):
    
    dataset = pd.read_parquet(working_folder + f"/data/{dataset_name}_data.parquet")
    #display(dataset)

    # Replace NaN with an empty string
    dataset['mod_lines'] = dataset['mod_lines'].fillna('')

    # Function to count the number of modified lines
    def count_mod_lines(mod_line):
        # Split by comma and filter out empty strings
        lines = [line for line in mod_line.split(',') if line]
        return len(lines)

    # Apply the function to each row and create a new column 'total_mod_lines'
    dataset['total_mod_lines'] = dataset['mod_lines'].apply(count_mod_lines)

    # Calculate the sum of all modified lines
    total_modified_lines = dataset['total_mod_lines'].sum()

    # Count the lines of code in each entry by counting '\n' and adding 1
    dataset['lines_of_code'] = dataset['code'].str.count('\n') + 1

    # Calculate the sum of lines of code for all entries
    total_lines_of_code = dataset['lines_of_code'].sum()

    print(dataset_name)
    print(f"Total unique project: {dataset['repo_url'].nunique()}")
    print(f'Total lines of code: {total_lines_of_code}')
    print(f'Total vul lines: {total_modified_lines}')
    print(f'Total non-vul lines:', total_lines_of_code-total_modified_lines)
    
show_information("Kotlin")
show_information("Swift")
show_information("Rust")

Kotlin
Total unique project: 4
Total lines of code: 1253
Total vul lines: 45
Total non-vul lines: 1208
Swift
Total unique project: 7
Total lines of code: 8195
Total vul lines: 104
Total non-vul lines: 8091
Rust
Total unique project: 19
Total lines of code: 26829
Total vul lines: 350
Total non-vul lines: 26479


kotlin project: 4, mod lines: 45
swift project 7, mod lines: 104
rust project 19m mod lines: 350

In [ ]:
DataUtils.read_language_dataset("kotlin", working_folder, parser)
DataUtils.read_language_dataset("swift", working_folder, parser)
DataUtils.read_language_dataset("rust", working_folder, parser)

In [6]:
swift_path = os.path.join(working_folder, f"data/swift/data_with_comments_single_before_change.csv")
swift_data = pd.read_csv(swift_path)

swift_data['processed_func'] = swift_data['processed_func'].apply(lambda code: DataUtils.remove_comments(parser, code, 'swift'))

output_path = os.path.join(working_folder, f"data/swift/data_single.csv")
swift_data.to_csv(output_path, index=False)

In [8]:
rust_path = os.path.join(working_folder, f"data/rust/data_with_comments_single_before_change.csv")
rust_data = pd.read_csv(rust_path)

rust_data['processed_func'] = rust_data['processed_func'].apply(lambda code: DataUtils.remove_comments(parser, code, 'rust'))

output_path = os.path.join(working_folder, f"data/rust/data_single.csv")
rust_data.to_csv(output_path, index=False)